### PR TITLE
IN with Enums numeric values

### PR DESCRIPTION
**Pull request by dimarub2000**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Backward Incompatible Change

Short description (up to few sentences):
Now numeric values for Enums can be used directly in IN section. #6766

### CODE REVIEW

File: dbms/src/Interpreters/convertFieldToType.cpp [hunk 1]
```diff
@@ -181,9 +181,16 @@ Field convertFieldToTypeImpl(const Field & src, const IDataType & type, const ID
         if (!which_type.isDateOrDateTime() && !which_type.isUUID() && !which_type.isEnum())
             throw Exception{"Logical error: unknown numeric type " + type.getName(), ErrorCodes::LOGICAL_ERROR};
 
-        /// Numeric values for Enums should not be used directly in IN section
-        if (src.getType() == Field::Types::UInt64 && !which_type.isEnum())
+        if (src.getType() == Field::Types::UInt64)
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/convertFieldToType.cpp]: What about Int64?

Comment (dimarub2000) [reply dbms/src/Interpreters/convertFieldToType.cpp]: Ok, but not sure if returning Int64 in else section won't break anything.

**Comment by alexey-milovidov** [comment]
Test needed.

File: dbms/src/Interpreters/convertFieldToType.cpp [hunk 2]
```diff
@@ -181,8 +181,13 @@ Field convertFieldToTypeImpl(const Field & src, const IDataType & type, const ID
...
 
-        /// Numeric values for Enums should not be used directly in IN section
-        if (src.getType() == Field::Types::UInt64 && !which_type.isEnum())
+        if (which_type.isEnum() && (src.getType() == Field::Types::UInt64 || src.getType() == Field::Types::Int64))
+        {
+            /// Convert UInt64 or Int64 to Enum's value
+            return dynamic_cast<const IDataTypeEnum &>(type).castToValue(src);
+        }
+
+        if (src.getType() == Field::Types::UInt64)
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/convertFieldToType.cpp]: Why do we need this branch?

Comment (dimarub2000) [reply dbms/src/Interpreters/convertFieldToType.cpp]: Will try to investigate. Seems like my changes doesn't really help.

Comment (dimarub2000) [reply dbms/src/Interpreters/convertFieldToType.cpp]: UPD: It does work the same as String now.
I was surprised with
```
CREATE TABLE enums (x Enum8('hello' = 0, 'world' = 1), y String) engine = Memory;
insert into enums values ('hello', 'find me'), ('hello', 'and me');
select y in ('hello') from enums;
┌─in(y, 'hello')─┐
│              0 │
│              0 │
└────────────────┘
select y in ('world') from enums;
┌─in(y, 'world')─┐
│              0 │
│              0 │
└────────────────┘

```

Comment (alexey-milovidov) [reply dbms/src/Interpreters/convertFieldToType.cpp]: Ok. And the only remaining question: do we still need this branch?

Comment (dimarub2000) [reply dbms/src/Interpreters/convertFieldToType.cpp]: The only problem i have found is in table function values :). When we do `const TupleBackend & value_tuple = value_field.safeGet<Tuple>().toUnderType();` to iterate the tuple values. All the fileds also go to under type and Data becomes UInt64.

Comment (dimarub2000) [reply dbms/src/Interpreters/convertFieldToType.cpp]: I believe that branch is not needed anymore. But not sure 100%. Deleting It also helped me to find bug in table function values.

**Comment by dimarub2000** [comment]
Ok, `Expected: Date. Got: UInt64` is a common thing.

**Pull request merged**